File: remote-gui/src/remote/protocol/clientserver/ClientMsg.java
package remote.protocol.clientserver;

import java.io.DataInput;
import java.io.DataOutput;
import remote.protocol.MsgIO;
import remote.protocol.MsgUint32;

public class ClientMsg extends MsgIO {
	private static final long currentProtocolVersion = 1;
	private MsgUint32 protocolVersion = new MsgUint32(1);
	private ClientMsgType type = new ClientMsgType();
	private MsgSession session = new MsgSession();
	private MsgClientRequest clientRequest = new MsgClientRequest();
	private MsgClientConfirm clientConfirm = new MsgClientConfirm();

	public ClientMsg() {
		super();
	}

	public void read(DataInput is) throws Exception
	{
		protocolVersion.read(is);
		type.read(is);
		switch(type.getValue())
		{
			case ClientMsgType.SESSION:
				session.read(is);
				break;
			case ClientMsgType.CLIENTCONFIRM:
				clientConfirm.read(is);
				break;
			case ClientMsgType.CLIENTREQUEST:
				clientRequest.read(is);
				break;
			default:
				break;
		}
	}

	public void write(DataOutput os) throws Exception
	{
		protocolVersion.setValue(ClientMsg.currentProtocolVersion);
		protocolVersion.write(os);
		type.write(os);
		switch(type.getValue())
		{
			case ClientMsgType.SESSION:
				session.write(os);
				break;
			case ClientMsgType.CLIENTCONFIRM:
				clientConfirm.write(os);
				break;
			case ClientMsgType.CLIENTREQUEST:
				clientRequest.write(os);
				break;
			default:
				break;
		}
	}

	public MsgSession getSession() {
		return session;
	}

	public MsgSession getAuthConfirm() {
		return session;
	}

	public MsgUint32 getProtocolVersion() {
		return protocolVersion;
	}

	public MsgClientConfirm getClientConfirm() {
		return clientConfirm;
	}

	public MsgClientRequest getClientRequest() {
		return clientRequest;
	}

	public ClientMsgType getType() {
		return type;
	}


}

File: trunk/remote-mci/protocols/client_server/ClientMsg.cc
#include "ClientMsg.h"

namespace remote { namespace protocols { namespace client_server {

ClientMsg::ClientMsg(uint8_t*& buffer, uint32_t& buflen)
{
	buffer = this->read(buffer,buflen);
	deleteMsg = true;
}

ClientMsg::ClientMsg(MsgSession& message)
          : type(CLIENTMSG_SESSION),
            message(&message),
            deleteMsg(false)
{}

ClientMsg::ClientMsg(MsgClientRequest& message)
         : type(CLIENTMSG_CLIENTREQUEST),
            message(&message),
            deleteMsg(false)
{}

ClientMsg::ClientMsg(MsgClientConfirm& message)
         : type(CLIENTMSG_CLIENTCONFIRM),
            message(&message),
            deleteMsg(false)
{}

ClientMsg::~ClientMsg()
{
	if (deleteMsg) { delete this->message; }
}

uint32_t ClientMsg::getLength()
{
	return sizeof(protocolVersion) + sizeof(type) + message->getLength();
}

uint8_t* ClientMsg::write(uint8_t* buffer, uint32_t& buflen)
{
	protocolVersion = CurrentProtocolVersion;
	buffer = writevalue(protocolVersion,buffer,buflen);
	buffer = writevalue(type,buffer,buflen);
	buffer = message->write(buffer,buflen);
	return buffer;
}

uint8_t* ClientMsg::read(uint8_t* buffer, uint32_t& buflen)
{
	buffer = readvalue(protocolVersion,buffer,buflen);
	buffer = readvalue(type,buffer,buflen);
	switch (type)
	{
		case CLIENTMSG_SESSION:
			message = new MsgSession(buffer,buflen);
			break;
		case CLIENTMSG_CLIENTREQUEST:
			message = new MsgClientRequest(buffer,buflen);
			break;
		case CLIENTMSG_CLIENTCONFIRM:
			message = new MsgClientConfirm(buffer,buflen);
			break;
		default:
			__THROW__ ("Invalid serverMsg");
			break;
	}
	return buffer;
}

void ClientMsg::print(FILE* s)
{
	fprintf(s,"MESSAGE ServerMsg\n");
	fprintf(s,"protocolVersion: %u\n",protocolVersion);
	fprintf(s,"type: ");
	switch (type)
	{
		case CLIENTMSG_CLIENTREQUEST:
			fprintf(s,"SERVERMSG_SERVERREQUEST\n");
			break;
		case CLIENTMSG_CLIENTCONFIRM:
			fprintf(s,"SERVERMSG_SERVERCONFIRM\n");
			break;
		case CLIENTMSG_SESSION:
			fprintf(s,"SERVERMSG_SESSION\n");
			break;
		default:
			__THROW__ ("Invalid serverMsg");
			break;
	}
	message->print(s);
}

uint8_t ClientMsg::getType()
{
	return type;
}

uint32_t ClientMsg::getProtocolVersion()
{
	return protocolVersion;
}

MsgClientRequest& ClientMsg::getClientRequest()
{
	if (type != CLIENTMSG_CLIENTREQUEST)
		__THROW__ ("Cannot get ServerRequest when type is not SERVERMSG_SERVERREQUEST!");
	return *((MsgClientRequest*)message);
}

MsgClientConfirm& ClientMsg::getClientConfirm()
{
	if (type != CLIENTMSG_CLIENTCONFIRM)
		__THROW__ ("Cannot get ServerConfirm when type is not SERVERMSG_SERVERCONFIRM!");
	return *((MsgClientConfirm*)message);
}

MsgSession& ClientMsg::getSession()
{
	if (type != CLIENTMSG_SESSION)
		__THROW__ ("Cannot getSession when command is not SERVERMSG_SESSION!");
	return *((MsgSession*)message);
}
}}}
